### PR TITLE
[RUNTIME] Add min_repeat_ms to time_evaluator

### PR DESCRIPTION
**Pull request by merrymercy**
`min_repeat_ms` sets the minimum duration of a measurement and has been used in autotvm for measurement.
As it is a useful feature to make measurement accurate and smart, we'd better move it to general API `time_evaluator` and encourage people to use it.

cc @eqy @tqchen   @sgrechanik-h

### CODE REVIEW

File: src/runtime/rpc/rpc_session.cc [hunk 1]
```diff
@@ -1190,21 +1192,39 @@ void RPCSession::EventHandler::HandlePackedCall() {
...
+PackedFunc WrapTimeEvaluator(PackedFunc pf,
+                             TVMContext ctx,
+                             int number,
+                             int repeat,
+                             int min_repeat_ms) {
+  auto ftimer = [pf, ctx, &number, repeat, min_repeat_ms](TVMArgs args, TVMRetValue *rv) {
     TVMRetValue temp;
     std::ostringstream os;
     // skip first time call, to activate lazy compilation components.
     pf.CallPacked(args, &temp);
```
Comment (eqy) [inline src/runtime/rpc/rpc_session.cc]: I wonder if this definition (1 + number * repeat) is the correct formulation after we have introduced `min_repeat_ms`. The goal is to start measurement in the correct power state, which we will likely do if we bump up `number` over and over again for the same `time_evaluator` call. However, let's say that `number` is now sufficient and we get to a fresh `time_evaluator` call. In this case I am not sure `1+` will be enough to get the hardware into the right state if necessary. Should we consider `number*(1+repeat)`?

File: src/runtime/rpc/rpc_session.cc [hunk 2]
```diff
@@ -1190,21 +1192,39 @@ void RPCSession::EventHandler::HandlePackedCall() {
...
+        }
+        DeviceAPI::Get(ctx)->StreamSync(ctx, nullptr);
+        tend = std::chrono::high_resolution_clock::now();
+
+        duration_ms = std::chrono::duration_cast<std::chrono::duration<double> >
+            (tend - tbegin).count() * 100;
+
+        if (duration_ms < min_repeat_ms) {
+          number = static_cast<int>(std::max((min_repeat_ms / (duration_ms / number) + 1),
+                                              number * 1.618));
```
[the remark this answers is not in the document]
Comment (eqy) [reply src/runtime/rpc/rpc_session.cc]: > What is `1.618`?
> Also, using ceil here might be better than adding 1.

https://en.wikipedia.org/wiki/Golden_ratio

File: src/runtime/rpc/rpc_module.cc [hunk 3]
```diff
@@ -124,7 +124,8 @@ class RPCModuleNode final : public ModuleNode {
   PackedFunc GetTimeEvaluator(const std::string& name,
                               TVMContext ctx,
                               int number,
-                              int repeat) {
+                              int repeat,
+                              int min_repeat_ms) {
```
Comment (eqy) [inline src/runtime/rpc/rpc_module.cc]: Does this break some current tests if we do not give a default value for `min_repeat_ms`?

**Comment by tqchen** [comment]
@merrymercy what is the status of this PR?

File: src/runtime/rpc/rpc_session.cc (hunk 1 above)
Comment (merrymercy) [reply src/runtime/rpc/rpc_session.cc]: Yes, the definition is not correct. I will add a note to the doc string of `min_repeat_ms` but keep this definition here.

File: src/runtime/rpc/rpc_session.cc [hunk 4]
```diff
@@ -1190,21 +1192,39 @@ void RPCSession::EventHandler::HandlePackedCall() {
...
+        // start timing
+        for (int i = 0; i < number; ++i) {
+          pf.CallPacked(args, &temp);
+        }
+        DeviceAPI::Get(ctx)->StreamSync(ctx, nullptr);
+        tend = std::chrono::high_resolution_clock::now();
+
+        duration_ms = std::chrono::duration_cast<std::chrono::duration<double> >
+            (tend - tbegin).count() * 100;
+
```
[the remark this answers is not in the document]
Comment (merrymercy) [reply src/runtime/rpc/rpc_session.cc]: @sgrechanik-h  We cannot use the accumulation mode due to the reason explained by eqy

File: src/runtime/rpc/rpc_session.cc (hunk 2 above)
[the remark this answers is not in the document]
Comment (merrymercy) [reply src/runtime/rpc/rpc_session.cc]: Precision is not very important here as I want to encourage it to set a higher `number`.

File: src/runtime/rpc/rpc_session.cc [hunk 5]
```diff
@@ -1190,21 +1192,39 @@ void RPCSession::EventHandler::HandlePackedCall() {
...
 }
 
-PackedFunc WrapTimeEvaluator(PackedFunc pf, TVMContext ctx, int number, int repeat) {
-  auto ftimer = [pf, ctx, number, repeat](TVMArgs args, TVMRetValue *rv) {
+PackedFunc WrapTimeEvaluator(PackedFunc pf,
+                             TVMContext ctx,
+                             int number,
+                             int repeat,
+                             int min_repeat_ms) {
+  auto ftimer = [pf, ctx, &number, repeat, min_repeat_ms](TVMArgs args, TVMRetValue *rv) {
```
[the remark this answers is not in the document]
Comment (merrymercy) [reply src/runtime/rpc/rpc_session.cc]: It turns out tvm packed function does not support capturing reference. I updated to capture by value.

File: src/runtime/rpc/rpc_module.cc (hunk 3 above)
Comment (merrymercy) [reply src/runtime/rpc/rpc_module.cc]: I added a default argument to python side

File: src/runtime/rpc/rpc_session.cc [hunk 6]
```diff
@@ -1190,23 +1192,42 @@ void RPCSession::EventHandler::HandlePackedCall() {
...
+                             int number,
+                             int repeat,
+                             int min_repeat_ms) {
+  auto ftimer = [pf, ctx, number, repeat, min_repeat_ms](TVMArgs args, TVMRetValue *rv) {
     TVMRetValue temp;
     std::ostringstream os;
     // skip first time call, to activate lazy compilation components.
     pf.CallPacked(args, &temp);
     DeviceAPI::Get(ctx)->StreamSync(ctx, nullptr);
+    int dynamic_number = number;
```
Comment (sgrechanik-h) [inline src/runtime/rpc/rpc_session.cc]: You used to modify `number` directly which had a nice property of remembering the suitable value of `number` between runs. I think you can still achieve this effect by declaring the lambda as `mutable` (won't be thread-safe though, so I'm not sure).

File: src/runtime/rpc/rpc_session.cc [hunk 7]
```diff
@@ -1190,23 +1192,42 @@ void RPCSession::EventHandler::HandlePackedCall() {
...
+        }
+        DeviceAPI::Get(ctx)->StreamSync(ctx, nullptr);
+        tend = std::chrono::high_resolution_clock::now();
+
+        duration_ms = std::chrono::duration_cast<std::chrono::duration<double> >
+            (tend - tbegin).count() * 1000;
+
+        dynamic_number = static_cast<int>(
+            std::max((min_repeat_ms / (duration_ms / dynamic_number) + 1),
+                     dynamic_number * 1.618));
```
Comment (sgrechanik-h) [inline src/runtime/rpc/rpc_session.cc]: The choice of the constant needs an explanation inside the code.

Comment (merrymercy) [reply src/runtime/rpc/rpc_session.cc]: Halide uses 2 but I think there is no "correct" number, so it is a random number.

File: src/runtime/rpc/rpc_session.cc (hunk 6 above)
Comment (merrymercy) [reply src/runtime/rpc/rpc_session.cc]: Sorry that my previous conclusion about tvm packed func is wrong. It works well with vanilla capture-by-reference. So I turn it back and tests are also added.

**Comment by merrymercy** [comment]
@eqy please review again

**Comment by tqchen** [comment]
ping @eqy @sgrechanik-h please take another look, if there is no further comments in 24 hours, we can go ahead and merge this PR in

**Comment by tqchen** [comment]
Thanks, @merrymercy @eqy @sgrechanik-h , this is merged

**Pull request merged**